### PR TITLE
Let driver own pcollections

### LINKED ISSUE #34: Refactor how we manage pcollections
Currently all the data (PCollections) is managed by PcollectionBufferManager actor and `ray.put()` into object store in `ray_execute_bundle` worker process. This approach has several implications:

1.  The owner of the data is whoever calls `ray.put()`, in our case, the worker processes that execute `ray_execute_bundle`. This is not ideal for fault tolerance since the data is fate share with the owner. In our cases, the data fate share with all the worker processes. Instead, we want the driver program to be the owner of all the data. This means instead of doing `ray.put()` we should return data as part of `ray_execute_bundle` return values so the driver program becomes the owner. More on the ownership design: https://docs.google.com/document/d/1lAy0Owi-vPz2jEqBSaHNQcy2IBSDEHyXNOQZlGuj93c/preview#heading=h.vjc9egi2q5aa
2. Currently `ray_execute_bundle` gets the input data references from PcollectionBufferManager and call `ray.get()` to fetch the actual data. This way ray scheduler cannot do locality aware scheduling of `ray_execute_bundle` remote function since by examining the function arguments, it doesn't know what data it needs. Instead, we should pass data ObjRefs directly as arguments to `ray_execute_bundle`.

At high level, it should look like:
```
output_data_obj_ref1, output_data_obj_ref2, ... = ray_execute_bundle.remote(input_data_obj_ref1, input_data_obj_ref2...)

ray_execute_bundle(input_data_obj_1, input_data_obj_2...):
   return output_data_obj_1, output_data_obj_2, ...
```

### PR DESCRIPTION
**Pull request by jjyao**
Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

With this PR, driver will own all the data of pcollections and handle the object lose automatically.

Closes #34

### CODE REVIEW

**Review by iasoon — APPROVED** [review]
Very nice @jjyao!

As the reason for this packing/unpacking of the arguments is a bit non-obvious in my opinion, I'm wondering what the best way would be to make these kinds of design decisions discoverable to readers of the code. Any thoughts? (Certainly not blocking to merge this)

File: ray_beam_runner/portability/execution.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def ray_execute_bundle(
     stage_timers: Mapping[translations.TimerFamilyId, bytes],
     instruction_request_repr: Mapping[str, typing.Any],
     dry_run=False,
-) -> Tuple[str, List[Any], Mapping[str, ray.ObjectRef]]:
+) -> Tuple[str, Any, Any]:  # (serialized InstructionResponse, repeat of pcoll, data)
```
[the remark this answers is not in the document]
Comment (pabloem) [reply ray_beam_runner/portability/execution.py:63]:
```suggestion
) -> Tuple[str, ...]:  # (serialized InstructionResponse, repeat of pcoll, data)
```
Like this?

**Comment by pabloem** [comment]
thanks everyone : )

**Comment by jjyao** [comment]
@iasoon, sorry for the late reply.

I'll apply https://github.com/ray-project/ray/pull/28291 to this PR so we no longer need to calculate `num_returns` beforehand.

**Pull request merged**